charon/src/llbc_ast_utils.rs

```diff
@@ -1,15 +1,84 @@
 //! Implementations for llbc_ast.rs
 
 #![allow(dead_code)]
-use crate::common::*;
+use crate::expressions::{Operand, Place, Rvalue};
 use crate::formatter::Formatter;
-use crate::im_ast::{fmt_call, FunDeclId, FunSigFormatter, GAstFormatter, TAB_INCR};
-use crate::llbc_ast::{Call, FunDecl, FunDecls, Statement, SwitchTargets};
+use crate::im_ast::{fmt_call, FunDeclId, FunSigFormatter, GAstFormatter, GlobalDeclId, TAB_INCR};
+use crate::llbc_ast::{Call, FunDecl, FunDecls, GlobalDecl, GlobalDecls, Statement, SwitchTargets};
 use crate::types::*;
 use crate::values::*;
+use crate::{common::*, id_vector};
+use itertools::chain;
 use serde::ser::SerializeTupleVariant;
 use serde::{Serialize, Serializer};
 
+impl Default for Statement {
+    fn default() -> Statement {
+        Statement::Nop
+    }
+}
+
+/// Goes from e.g. [A, B, C] ; D to (A, (B, (C, D))).
+pub fn chain_statements(binds: Vec<Statement>, last: Statement) -> Statement {
+    binds.into_iter().rev().fold(last, |cont, bind| {
+        Statement::Sequence(Box::new(bind), Box::new(cont))
+    })
+}
+
+/// Visit an rvalue and generate statements. Used below in [transform_operands].
+fn transform_rvalue_operands<F: FnMut(&mut Operand) -> Vec<Statement>>(
+    rval: &mut Rvalue,
+    f: &mut F,
+) -> Vec<Statement> {
+    match rval {
+        Rvalue::Use(op) | Rvalue::UnaryOp(_, op) => f(op),
+        Rvalue::BinaryOp(_, o1, o2) => chain(f(o1), f(o2)).collect(),
+        Rvalue::Aggregate(_, ops) => ops.iter_mut().flat_map(f).collect(),
+        Rvalue::Discriminant(_) | Rvalue::Ref(_, _) => vec![],
+    }
+}
+
+/// Map a statement to the given visitor of operands while inserting the generated statements.
+/// Useful to implement passes on operands.
+pub fn transform_operands<F: FnMut(&mut Operand) -> Vec<Statement>>(
+    st: Statement,
+    f: &mut F,
+) -> Statement {
+    // Does two matchs, depending if we want to move or to borrow the statement.
```

Comment:
Consuming statements works best when I want to change their structure or consume their insides, but it's more painful when I just want to do in-place update (e.g. `a.b.c += 1` vs `a = {... b = {... .c = old.c + 1}}`). This is what happens here, where I just want to update in-place the operands, then consume the whole statement with `chain_statements` (you can see that I consume `st` in the `match &mut` branches). But I would not be able to consume the whole statement statement if I was consuming it in the match.

Reply:
Concerning the callback signature, I prefer having the new statements in the output instead of putting them in a `&mut` argument (in general, outputs should be IMO in the return type instead of a mutable parameter if possible). Then, about taking & returning the operand instead of modifying it, I have no strong opinion, the issue being here that I cannot consume it if I match the statement by `&mut` (and if I don't do that as said above, I can't consume the whole statement in the `match` branches).

Reply:
I don't know if it's clearer ...

Reply:
Let's discuss that orally

Reply:
I extracted the `chain_statements` so it's a bit more clear